File: sequential.py
"""-----------------------------------------------------------------------------
 "
 " File:          sequential.py
 " Author:        Cody Walker
 " Project:       Parallel Project 1 - n-Bodies and Collisions
 " Description:   A sequential implementation of a simulation of the n-body
 "                gravitational problem written in Python in order to run a
 "                time test between implementations written in different
 "                languages.
 " Created:       12 June 2019
 " Updated:       12 June 2019
 "
 " NOTES:
 "              UNIQUE  |   ORDER   |   CHANGE SPEC EL  |   NEW ELS
 "  LISTS         N           Y                Y               Y
 "  TUPLES        N           Y                N               N
 "  DICT          Y           N                Y               Y
 "  SETS          Y           N                N               Y
 "
 " --------------------------------------------------------------------------"""

            # TODO: SPEED UP WITH DISTANCES LIST OR SUM10


import math
import time

# Test for gcommit


"""--------------------------------------------------------------------------"""
"""------------------------------- Globals ----------------------------------"""
"""--------------------------------------------------------------------------"""


pdb = 1     # Debug print statements.

NUM_BODIES = 4          # Number of bodies in the simulation.
TIMESTEPS = 30          # Number of timesteps to be run in simulation.
BODY_RADIUS =  5        # Radius of each body in the simulation.
BODY_MASS = 100000000   # Mass of each body in the simulation.
ENABLE_GUI = 0          # If 1, write coords to file for visual simulation.

GRAV_CONST = 6674.08    # G scaled by 1000 to make numbers easier to work with.
SPECIAL_G = 2 * GRAV_CONST * BODY_MASS  # To lessen number of computations.

xposition = []  # List of x-positions of the bodies.
yposition = []  # List of y-positions of the bodies.
xvelocity = []  # List of x-velocities of the bodies.
yvelocity = []  # List of y-velocities of the bodies.
xforce = []     # List of net x-forces acting upon each of the bodies.
yforce = []     # List of net y-forces acting upon each of the bodies.

collisions = [] # List of collisions that occurred in current timestep.


"""--------------------------------------------------------------------------"""
"""------------------------------ Functions ---------------------------------"""
"""--------------------------------------------------------------------------"""


"""-----------------------------------------------------------------------------
 " Function:    init
 " Description: Initializes global variables, setting them to be any user-
 "              supplied values if appropriate.
 " Arguments:   argv1 - number of bodies
 "              argv2 - radius of each body
 "              argv3 - number of time steps
 "              argv4 - 1 to enable output for gui, 0 otherwise
 " Returns:     Nothing.
 " --------------------------------------------------------------------------"""
def init():

    # TODO: take command line args?

    # Open file in write mode in order to start fresh.
    with open("testie.txt", "w") as file:
        # Write num_bodies, body_radius, & timesteps to file for GUI.
        file.write(str(NUM_BODIES))
        file.write(" ")
        file.write(str(BODY_RADIUS))
        file.write(" ")
        file.write(str(TIMESTEPS))
        file.write(" ")

    # Close the file.
    file.close()

    # Initialize x & y positions.
    # TODO: ACTUALLY MAKE DIS RANDOM LATER
    for i in range(NUM_BODIES):
        xposition.append(i * 100)
        yposition.append(i * 100)

    # Initialize x & y velocities & forces as 0.
    for _ in range(NUM_BODIES):
        xvelocity.append(0)
        yvelocity.append(0)
        xforce.append(0)
        yforce.append(0)

    return
    """ END init() """


"""-----------------------------------------------------------------------------
 " Function:    calculate_forces
 " Description: For each body, calculates the net force acting upon it from all
 "              the other bodies in the simulation in the current timeframe.
 "              The function does not move any of the bodies, and merely does
 "              the calculations for all the forces at the current moment in
 "              time.
 " Arguments:   None.
 " Returns:     Nothing.
 " --------------------------------------------------------------------------"""
def calculate_forces():

    for i in range(NUM_BODIES - 1):
        # Start j at i + i since we save the forces to BOTH bodies each iter.
        for j in range(i + 1, NUM_BODIES):
            # Calculate distance between current pair of bodies.
            x_dist_chunk = (xposition[i] - xposition[j]) * (xposition[i] -
                                                            xposition[j])
            y_dist_chunk = (yposition[i] - yposition[j]) * (yposition[i] -
                                                            yposition[j])
            distance = math.sqrt(x_dist_chunk + y_dist_chunk)
            # Calculate magnitude of gravitational force between pair of bodies
            magnitude = SPECIAL_G / (distance * distance)
            # Precompute some shunks to lessen total # of computations.
            xdirection = xposition[j] - xposition[i]
            ydirection = yposition[j] - yposition[i]
            mag_over_dist = (magnitude / distance)
            x_chunk = (mag_over_dist * xdirection)
            y_chunk = (mag_over_dist * ydirection)
            # Save BOTH forces to cut calculations in half.
            xforce[i] = xforce[i] + x_chunk
            xforce[j] = xforce[j] - x_chunk
            yforce[i] = yforce[i] + y_chunk
            yforce[j] = yforce[j] - y_chunk
    return
    """ END calculate_forces() """


"""-----------------------------------------------------------------------------
 " Function:    move_bodies
 " Description: For each body, determines first its new velocity as a result of
 "              the net forct acting upon it, and then its new position as a
 "              result of its new velocity.
 " Arguments:   None.
 " Returns:     Nothing.
 " --------------------------------------------------------------------------"""
def move_bodies():
    for i in range(NUM_BODIES):
        # Calculate change in velocity resulting from net force on current obj.
        x_delta_v = xforce[i] / BODY_MASS
        y_delta_v = yforce[i] / BODY_MASS
        # Calculate change in position resulting from curr object's velocity.
        x_delta_p = xvelocity[i] + (x_delta_v * 0.5)
        y_delta_p = yvelocity[i] + (y_delta_v * 0.5)
        # Store current object's new velocity.
        xvelocity[i] = xvelocity[i] + x_delta_v
        yvelocity[i] = yvelocity[i] + y_delta_v
        # Store current object's new position.
        xposition[i] = xposition[i] + x_delta_p
        yposition[i] = yposition[i] + y_delta_p
        # Reset force vector to prepare for next time move_bodies called.
        xforce[i] = 0
        yforce[i] = 0
    return
    """ END move_bodies() """


"""-----------------------------------------------------------------------------
 " Function:    collisions_detected
 " Description: Checks the current state of the simulation, and determines
 "              whether or not the previous movement calculation has caused any
 "              collisions to occur.
 " Arguments:   None.
 " Returns:     1 if any collisions have occured, 0 otherwise.
 " --------------------------------------------------------------------------"""
def collisions_detected():
    is_collision = 0    # Boolean representing if collision occurred.
    for i in range(NUM_BODIES - 1):
        for j in range(i + 1, NUM_BODIES):
            # TODO: SPEED UP WITH DISTANCES LIST OR SUM10
            # Calculate distance between current pair of bodies.
            x_dist_chunk = (xposition[i] - xposition[j]) * (xposition[i] -
                                                            xposition[j])
            y_dist_chunk = (yposition[i] - yposition[j]) * (yposition[i] -
                                                            yposition[j])
            distance = math.sqrt(x_dist_chunk + y_dist_chunk)
            if (distance < (2 * BODY_RADIUS)):
                if (pdb):
                    print("\nCOLLISION!!!!!!!\n")
                # Mark that these two objects in particular collided.
                collisions.append((i,j))
                is_collision = 1
    return is_collision
    """ END collisions_detected() """


"""-----------------------------------------------------------------------------
 " Function:    resolve_collisions
 " Description: Called when two or more objects are in a "state of collision"
 "              to properly determine their new velocities and positions
 "              resulting from them colliding.  Without this method, not only
 "              would objects pass through eachother, but because the
 "              magnitude of the force of gravity becomes exponentially larger
 "              the closer together bodies get, the bodies would become under
 "              tremendous force and fly off at comical speeds in the direction
 "              they were just heading.  Very important function.
 "              NOTE:
 "              THE PROCEDURES USED TO CALCULATE THE FINAL VELOCITIES OF THE
 "              OBJECTS ARE TAKEN FROM DR. HOMER'S HANDOUT WHICH IS BADED ON
 "              "2-Dimensional Elastic Collisions Without Trigonometry" by
 "              Chad Bercheck, AND IS NOT MY OWN ORIGINAL MATERIAL.
 " Arguments:   None.
 " Returns:     Nothing.
 " --------------------------------------------------------------------------"""
def resolve_collisions():
    # Handle every collision that occurred in current timestep.
    while (len(collisions) > 0):
        # First, get the two bodies involved in current collision.
        curr = collisions.pop()
        body1 = curr[0]
        body2 = curr[1]
        if (pdb):
            print("body1:", body1, "  ", "body2:", body2)

        # Extract information about velocities and positions of bodies.
        x1 = xposition[body1]
        y1 = yposition[body1]
        x2 = xposition[body2]
        y2 = yposition[body2]
        v1x = xvelocity[body1]
        v1y = yvelocity[body1]
        v2x = xvelocity[body2]
        v2y = yvelocity[body2]

        # Precompute simm. chunks used in all equations - reduce computations.
        denom = ((x2 - x1) * (x2 - x1)) + ((y2 - y1) * (y2 - y1))
        chunk1 = (v1y * (x2 - x1) * (y2 - y1))
        chunk2 = (v2y * (x2 - x1) * (y2 - y1))
        chunk3 = (v1x * (y2 - y1) * (x2 - x1))
        chunk4 = (v2x * (x2 - x1) * (y2 - y1))

        # Determine final x velocity of first body.
        v1fx = v2x * (x2 - x1) * (x2 - x1)
        v1fx = v1fx + chunk2
        v1fx = v1fx + (v1x * (y2 - y1) * (y2 - y1))
        v1fx = v1fx - chunk1
        v1fx = v1fx / denom

        # Determine final y velocity of first body.
        v1fy = chunk4
        v1fy = v1fy + (v2y * (y2 - y1) * (y2 - y1))
        v1fy = v1fy - chunk3
        v1fy = v1fy + (v1y * (x2 - x1) * (x2 - x1))
        v1fy = v1fy / denom

        # Determine final x velocity of second body.
        v2fx = v1x * (x2 - x1) * (x2 - x1)
        v2fx = v2fx + chunk1
        v2fx = v2fx + (v2x * (y2 - y1) * (y2 - y1))
        v2fx = v2fx - chunk2
        v2fx = v2fx / denom

        # Determine final y velocity of second body.
        v2fy = chunk3
        v2fy = v2fy + (v1y * (y2 - y1) * (y2 - y1))
        v2fy = v2fy - chunk4
        v2fy = v2fy + (v2y * (x2 - x1) * (x2 - x1))
        v2fy = v2fy / denom

        # Determine final positions of bodies.
        xposition[body1] = xposition[body1] + v1fx
        yposition[body1] = yposition[body1] + v1fy
        xposition[body2] = xposition[body2] + v2fx
        yposition[body2] = yposition[body2] + v2fy

        # Copy new velocities into global lists.
        xvelocity[body1] = v1fx
        yvelocity[body1] = v1fy
        xvelocity[body2] = v2fx
        yvelocity[body2] = v2fy

    return
    """ END resolve_collisions() """


"""-----------------------------------------------------------------------------
 " Function:    export_positions
 " Description: Writes the information contained inside the position array at
 "              the current timeframe to a .txt file for a GUI to read and
 "              use to generate a visualization of the simulation.
 " Arguments:   None.
 " Returns:     Nothing.
 " --------------------------------------------------------------------------"""
def export_positions():
    # TODO: IMPLEMENT THIS METHOD
    return
    """ END export_positions() """


"""-----------------------------------------------------------------------------
 " Function:    print_coordinates
 " Description: Prints out the position coordinates of every body in the
 "              simulation.
 " Arguments:   None.
 " Returns:     Nothing.
 " --------------------------------------------------------------------------"""
def print_coordinates():
    print("\nPOSITIONS:")
    for i in range(NUM_BODIES):
        print("Body", i, "x =", xposition[i], "y =", yposition[i])
    return
    """ END print_coordinates() """


"""-----------------------------------------------------------------------------
 " Function:    print_velocities
 " Description: Prints out the velocities of every body in the simulation.
 " Arguments:   None.
 " Returns:     Nothing.
 " --------------------------------------------------------------------------"""
def print_velocities():
    print("VELOCITIES:\n")
    for i in range(NUM_BODIES):
        print("Body", i, "vx =", xvelocity[i], "vy =", yvelocity[i])
    return
    """ END print_velocities() """


"""-----------------------------------------------------------------------------
 " Function:    print_forces
 " Description: Prints out the net forces acting upon every body in the
 "              simulation.
 " Arguments:   None.
 " Returns:     Nothing.
 " --------------------------------------------------------------------------"""
def print_forces():
    print("FORCES:\n")
    for i in range(NUM_BODIES):
        print("Body", i, "x =", xposition[i], "y =", yposition[i])
    return
    """ END print_forces() """


"""-----------------------------------------------------------------------------
 " Function:    print_collisions
 " Description: Prints out the tuples that are currently in the collision list.
 " Arguments:   None.
 " Returns:     Nothing.
 " --------------------------------------------------------------------------"""
def print_collisions():
    print("COLLISIONS:")
    for collision in collisions:
        print("COLLISION between bodies: ", collision)
    return
    """ END print_collisions() """


"""-----------------------------------------------------------------------------
 " Function:    main
 " Description: Doesn't do much, just calls functions to do all the work
 "              instead.  Records system time before and after doing the work
 "              to be used for measurement of execution time.
 " Arguments:   argv1 - number of bodies
 "              argv2 - radius of each body
 "              argv3 - number of time steps
 "              argv4 - 1 to enable output for gui, 0 otherwise
 "              argv5 - number of dimensions to run simulation in (2 or 3)
 " Returns:     Appropriate exit status of program.
 " --------------------------------------------------------------------------"""
def main():

    # TODO: Time calculation bs

    start_time = time.perf_counter()

    init()
    if (pdb):
        print_coordinates()

    # Run Simulation for # of timesteps requested.
    for i in range(TIMESTEPS):

        # Calculate net force acting on each body.
        calculate_forces()

        # Move bodies appropriately based upon the net force acting upon them.
        move_bodies()

        # Check for collisions and resolve any that are found.
        if (collisions_detected()):
            if(pdb):
                print_collisions()
            resolve_collisions()
            if(pdb):
                print_collisions()

        # If print debug variable turned on, print to stdout.
        if (pdb):
            print_coordinates()

        # If GUI mode enabled write positions to file.
        if (ENABLE_GUI):
            export_positions()

    end_time = time.perf_counter()
    print("\n\nTOTAL TIME ELAPSED:   ", end_time - start_time, "\n\n")

    return
    """ END main() """


# The call to main() that gets the whole simulation started.
main()


"""-----------------------------------------------------------------------------
 " Function:    func
 " Description: desc
 " Arguments:   None.
 " Returns:     Nothing.
 " --------------------------------------------------------------------------"""

""" END sequential.py """

File: sequential.c
/* -----------------------------------------------------------------------------
 *
 * File:          sequential.c
 * Author:        Cody Walker
 * Class:         CSC 422 Parallel & Distributed Programming
 * Project:       Parallel Project 1 - n-Bodies and Collisions
 * Description:   A sequential implementation of a simulation of the n-body
 *                gravitational problem.
 * Created:       10 October 2018
 * Updated:       13 June 2019
 *
 * -------------------------------------------------------------------------- */


// TODO:    PEEP DISTANCE STRUCTURE!
// TODO:    change pow(x) to x * x wherever applicable.
// TODO:    MORE SIMILAR CHUNKS! i.e. (x2 - x1)^2 etc.


#include "sequential.h"

#define NUM_BODIES      50
#define TIMESTEPS       300
#define ENABLE_GUI      1
#define BODY_RADIUS     20
#define BODY_MASS       1000000000

int pdb = 1;


/* -------------------------------------------------------------------------- */
/* ----------------------------- Functions ---------------------------------- */
/* -------------------------------------------------------------------------- */



/* -----------------------------------------------------------------------------
 * Routine:     init
 * Description: Initializes global variables, setting them to be any user-
 *              supplied values if appropriate.
 * Arguments:   argv1 - number of bodies
 *              argv2 - radius of each body
 *              argv3 - number of time steps
 *              argv4 - 1 to enable output for gui, 0 otherwise
 * Returns:     Nothing.
 * -------------------------------------------------------------------------- */
void init(int argc, char *argv[])
{
  int i;

  // Initialize global values.
  num_bodies = NUM_BODIES;
  body_radius = BODY_RADIUS;
  timesteps = TIMESTEPS;
  enable_gui = ENABLE_GUI;

  body_mass = BODY_MASS;
  file = NULL;
  secs = 0;
  usecs = 0;

  // Pull values from arguments.
//  num_bodies = atoi(argv[1]);
//  body_radius = atoi(argv[2]);
//  timesteps = atoi(argv[3]);
//  enable_gui = atoi(argv[4]);

  // Open file.
  file = fopen("gui_input.txt", "w");

  // Write num_bodies, body_radius, & timesteps to file for GUI.
  if (fprintf(file, "%d ", num_bodies) == EOF)
  {
    error("ERROR!!! init(): PROBLEM WITH fprintf()");
  }
  if (fprintf(file, "%d ", body_radius) == EOF)
  {
    error("ERROR!!! init(): PROBLEM WITH fprintf()");
  }
  if (fprintf(file, "%d ", timesteps) == EOF)
  {
    error("ERROR!!! init(): PROBLEM WITH fprintf()");
  }

  // Set value for universal gravitational constant.
  grav_const = 6674.08;   // Scaled a bit so units are easier to work with.
  special_g = 2 * grav_const * body_mass;



  int randx = 0;
  int randy = 0;


  // Initialize positions.
  position = malloc(num_bodies * sizeof(vector *));
  for (i = 0; i < num_bodies; i++)
  {
    position[i] = malloc(sizeof(vector));
    // ACTUALLY INITIALIZE THEM SOMEHWERE USEFUL


    position[i]->x = i * 100;
    position[i]->y = i * 100;

    randx = rand() % 100;
    randy = rand() % 100;

    position[i]->x += randx;
    position[i]->y += randy;

 }

  // Initialize velocities.
  velocity = malloc(num_bodies * sizeof(vector *));
  for (i = 0; i < num_bodies; i++)
  {
    velocity[i] = malloc(sizeof(vector));
    velocity[i]->x = 0;
    velocity[i]->y = 0;
  }

  // Initialize forces.
  force = malloc(num_bodies * sizeof(vector *));
  for (i = 0; i < num_bodies; i++)
  {
    force[i] = malloc(sizeof(vector));
    force[i]->x = 0;
    force[i]->y = 0;
  }

  // Initialize collision queue.
  collisions.size = 0;
  collisions.head = NULL;
  collisions.tail = NULL;

  return;
} /* END init() */


/* -----------------------------------------------------------------------------
 * Routine:     terminate
 * Description: Cleans up all allocated memory.
 * Arguments:   None.
 * Returns:     Nothing.
 * -------------------------------------------------------------------------- */
void terminate(void)
{
  int i;

  // Retrieve values and print.
  secs = (end_tv.tv_sec - start_tv.tv_sec);
  usecs = (end_tv.tv_usec - start_tv.tv_usec);
  if (usecs < 0)
  {
    secs = secs - 1;
    usecs = usecs + 1000000;
  }

  printf("\n\nSIMULATION TOOK %ld.%ld SECONDS.\n\n", secs, usecs);

  // Free position array.
  for (i = 0; i < num_bodies; i++)
  {
    free(position[i]);
  }
  free(position);

  // Free velocity array.
  for (i = 0; i < num_bodies; i++)
  {
    free(velocity[i]);
  }
  free(velocity);

  // Free force array.
  for (i = 0; i < num_bodies; i++)
  {
    free(force[i]);
  }
  free(force);

  // Close file.
  fclose(file);

  return;
} /* END terminate() */


/* -----------------------------------------------------------------------------
 * Routine:     calculate_forces
 * Description: For each body, calculates the net-force acting upon it from all
 *              the other bodies in the simulation in the current timeframe.
 *              The function does not move any of the bodies, and merely does
 *              the calculations for all the forces at the current moment in
 *              time.
 * Arguments:   None.
 * Returns:     Nothing.
 * -------------------------------------------------------------------------- */
void calculate_forces(void)
{
  int i, j;
  double distance = 0.0;
  double magnitude = 0.0;
  double x_chunk = 0.0;
  double y_chunk = 0.0;
  double mag_over_dist = 0.0;
  vector direction;
  direction.x = 0.0;
  direction.y = 0.0;

  for (i = 0; i < num_bodies - 1; i++)
  {
    // Start j at i + 1 since we save the force to BOTH bodies each iteration.
    for (j = i + 1; j < num_bodies; j++)
    {
      // Calculate distance between current pair of bodies.
//      distance = sqrt(pow((position[i]->x - position[j]->x), 2) +
//          pow((position[i]->y - position[j]->y), 2));
      distance = sqrt(((position[i]->x - position[j]->x) * (position[i]->x -
                  position[j]->x)) + ((position[i]->y - position[j]->y) *
                  (position[i]->y - position[j]->y)));
      // Calculate magnitude of gravitational force between pair of bodies.
      magnitude = special_g / (distance * distance);
      direction.x = position[j]->x - position[i]->x;
      direction.y = position[j]->y - position[i]->y;
      // Precompute some chunks to lessen total # of computations.
      mag_over_dist = (magnitude / distance);
      x_chunk = (mag_over_dist * direction.x);
      y_chunk = (mag_over_dist * direction.y);
      // Save both forces to cut calculations in half.
      force[i]->x = force[i]->x + x_chunk;
      force[j]->x = force[j]->x - x_chunk;
      force[i]->y = force[i]->y + y_chunk;
      force[j]->y = force[j]->y - y_chunk;
    }
  }
  return;
} /* END calculate_forces() */


/* -----------------------------------------------------------------------------
 * Routine:     move_bodies
 * Description: For each body, determines first its new velocity as a result of
 *              the net-force acting upon it, and then its new position as a
 *              result of its new velocity.
 * Arguments:   None.
 * Returns:     Nothing.
 * -------------------------------------------------------------------------- */
void move_bodies(void)
{
  int i;
  vector delta_v, delta_p;
  delta_v.x = 0.0;
  delta_v.y = 0.0;
  delta_p.x = 0.0;
  delta_p.y = 0.0;

  for (i = 0; i < num_bodies; i++)
  {
    // Calculate change in velocity resulting from net force on current obj.
    delta_v.x = force[i]->x / body_mass;
    delta_v.y = force[i]->y / body_mass;
    // Calculate change in position resulting from current object's velocity.
    delta_p.x = velocity[i]->x + (delta_v.x * 0.5);
    delta_p.y = velocity[i]->y + (delta_v.y * 0.5);
    // Store current object's new velocity.
    velocity[i]->x = velocity[i]->x + delta_v.x;
    velocity[i]->y = velocity[i]->y + delta_v.y;
    // Store current object's new position.
    position[i]->x = position[i]->x + delta_p.x;
    position[i]->y = position[i]->y + delta_p.y;
    // Reset force vector to prepare for next time move_bodies() is called.
    force[i]->x = 0.0;
    force[i]->y = 0.0;
  }
  return;
} /* END move_bodies() */


/* -----------------------------------------------------------------------------
 * Routine:     collisions_detected
 * Description: Checks the current state of the simulation, and determines
 *              whether or not the previous movement calculation has caused any
 *              collisions to occur.
 * Arguments:   None.
 * Returns:     1 if any collisions have occured, 0 otherwise.
 * -------------------------------------------------------------------------- */
int collisions_detected(void)
{
  int i, j;
  int collision = 0;  // Boolean representing if collision occured.
  double distance = 0.0;

  for (i = 0; i < num_bodies - 1; i++)
  {
    for (j = i + 1; j < num_bodies; j++)
    {
        // TODO: SPEED THIS SHIT UP MAYBE DISTANCE STRUCTURE??????
      // Calculate distance between current pair of bodies.
      distance = sqrt(pow((position[i]->x - position[j]->x), 2) +
          pow((position[i]->y - position[j]->y), 2));
      // Determine if bodies are currently "overlapping".
      if (distance < (2 * body_radius))
      {
        if (pdb) printf("\n\n\nCOLLISION!!!!!!!\n\n\n");
        // Mark that these two objects in particular collided.
        add_collision(i, j);
        collision = 1;
      }
    }
  }
  // Returns TRUE if collision happened, FALSE otherwise.
  return collision;
} /* END collisions_detected() */


/* -----------------------------------------------------------------------------
 * Routine:     resolve_collisions
 * Description: Called when two or more objects are in a "state of collision"
 *              to properly determine their new velocities and positions
 *              resulting from them colliding.  Without this method, not only
 *              would objects pass through eachother, but because the
 *              magnitude of the force of gravity becomes exponentially larger
 *              the closer together bodies get, the bodies would become under
 *              tremendous force and fly off at comical speeds in the direction
 *              they were just heading.  Very important function.
 *              NOTE:
 *              THE PROCEDURES USED TO CALCULATE THE FINAL VELOCITIES OF THE
 *              OBJECTS ARE TAKEN FROM DR. HOMER'S HANDOUT WHICH IS BADED ON
 *              "2-Dimensional Elastic Collisions Without Trigonometry" by
 *              Chad Bercheck, AND IS NOT MY OWN ORIGINAL MATERIAL.
 * Arguments:   None.
 * Returns:     Nothing.
 * -------------------------------------------------------------------------- */
void resolve_collisions(void)
{
  c_node curr;          // Collision object to extract indices from.
  int body1 = -1;       // Index of first body involved in the collision.
  int body2 = -1;       // Index of second body involved in the collision.
  double x1 = 0.0;      // Body 1 initial x position.
  double y1 = 0.0;      // Body 1 initial y position.
  double x2 = 0.0;      // Body 2 initial x position.
  double y2 = 0.0;      // Body 2 initial y position.
  double v1x = 0.0;     // Body 1 initial x velocity.
  double v1y = 0.0;     // Body 1 initial y velocity.
  double v2x = 0.0;     // Body 2 initial x velocity.
  double v2y = 0.0;     // Body 2 initial y velocity.
  double v1fx = 0.0;    // Body 1 final x velocity.
  double v1fy = 0.0;    // Body 1 final y velocity.
  double v2fx = 0.0;    // Body 2 final x velocity.
  double v2fy = 0.0;    // Body 2 final y velocity.
  // Used to speed up computation.
  double denom = 0.0;
  double chunk1 = 0.0;
  double chunk2 = 0.0;
  double chunk3 = 0.0;
  double chunk4 = 0.0;

    // TODO: MORE SIMILAR CHUNKS!!! i.e. (x2 - x1)^2 etc.

  // Handle every collision that occured in current timestep.
  while (collisions.size > 0)
  {
    // First, get the two bodies involved in current collision.
    curr = pop_collision();
    body1 = curr.a;
    body2 = curr.b;

    // Extract information about velocities and positions of bodies.
    x1 = position[body1]->x;
    y1 = position[body1]->y;
    x2 = position[body2]->x;
    y2 = position[body2]->y;
    v1x = velocity[body1]->x;
    v1y = velocity[body1]->y;
    v2x = velocity[body2]->x;
    v2y = velocity[body2]->y;

    // Precompute similar "chunks" used in all equations to reduce computations.
//    denom = (pow((x2 - x1), 2) + pow((y2 - y1), 2));
    denom = ((x2 - x1) * (x2 - x1)) + ((y2 - y1) * (y2 - y1));
    chunk1 = (v1y * (x2 - x1) * (y2 - y1));
    chunk2 = (v2y * (x2 - x1) * (y2 - y1));
    chunk3 = (v1x * (y2 - y1) * (x2 - x1));
    chunk4 = (v2x * (x2 - x1) * (y2 - y1));

    // Determine final x velocity of first body.
//    v1fx = (v2x * pow((x2 - x1), 2));
    v1fx = v2x * (x2 - x1) * (x2 - x1);
    v1fx = v1fx + chunk2;
//    v1fx = v1fx + (v1x * pow((y2 - y1), 2));
    v1fx = v1fx + (v1x * (y2 - y1) * (y2 - y1));
    v1fx = v1fx - chunk1;
    v1fx = v1fx / denom;

    // Determine final y velocity of first body.
    v1fy = chunk4;
//    v1fy = v1fy + (v2y * pow((y2 - y1), 2));
    v1fy = v1fy + (v2y * (y2 - y1) * (y2 - y1));
    v1fy = v1fy - chunk3;
//    v1fy = v1fy + (v1y * pow((x2 - x1), 2));
    v1fy = v1fy + (v1y * (x2 - x1) * (x2 - x1));
    v1fy = v1fy / denom;

    // Determine final x velocity of second body.
//    v2fx = (v1x * pow((x2 - x1), 2));
    v2fx = v1x * (x2 - x1) * (x2 - x1);
    v2fx = v2fx + chunk1;
//    v2fx = v2fx + (v2x * pow((y2 - y1), 2));
    v2fx = v2fx + (v2x * (y2 - y1) * (y2 - y1));
    v2fx = v2fx - chunk2;
    v2fx = v2fx / denom;

    // Determine final y velocity of second body.
    v2fy = chunk3;
//    v2fy = v2fy + (v1y * pow((y2 - y1), 2));
    v2fy = v2fy + (v1y * (y2 - y1) * (y2 - y1));
    v2fy = v2fy - chunk4;
//    v2fy = v2fy + (v2y * pow((x2 - x1), 2));
    v2fy = v2fy + (v2y * (x2 - x1) * (x2 - x1));
    v2fy = v2fy / denom;

    // Determine final positions of bodies.
    position[body1]->x = position[body1]->x + v1fx;
    position[body1]->y = position[body1]->y + v1fy;
    position[body2]->x = position[body2]->x + v2fx;
    position[body2]->y = position[body2]->y + v2fy;

    // Copy new velocities into global arrays.
    velocity[body1]->x = v1fx;
    velocity[body1]->y = v1fy;
    velocity[body2]->x = v2fx;
    velocity[body2]->y = v2fy;
  }
  return;
} /* END resolve_collisions() */


/* -----------------------------------------------------------------------------
 * Routine:     export_positions
 * Description: Writes the information contained inside the position array at
 *              the current timeframe to a .txt file for a GUI to read and
 *              use to generate a visualization of the simulation.
 * Arguments:   None.
 * Returns:     Nothing.
 * -------------------------------------------------------------------------- */
void export_positions(void)
{
  int i;

  for (i = 0; i < num_bodies; i++)
  {
    if (fprintf(file, "%lf ", position[i]->x) == EOF)
    {
      error("ERROR!!! export_positions(): PROBLEM WITH fprintf()");
    }
    if (fprintf(file, "%lf ", position[i]->y) == EOF)
    {
      error("ERROR!!! export_positions(): PROBLEM WITH fprintf()");
    }
  }
  return;
} /* END export_positions() */


/* -----------------------------------------------------------------------------
 * Routine:     main
 * Description: Doesn't do much, just calls methods to do all the work instead.
 *              Records system time before and after doing the work however to
 *              be used for measurement of execution time.
 * Arguments:   argv1 - number of bodies
 *              argv2 - radius of each body
 *              argv3 - number of time steps
 *              argv4 - 1 to enable output for gui, 0 otherwise
 *              argv5 - number of dimensions to run simulation in (2 or 3)
 * Returns:     Appropriate exit status of program.
 * -------------------------------------------------------------------------- */
int main(int argc, char *argv[])
{
  int i;

  init(argc, argv);

  if (pdb) print_coordinates();

  // Record system's time immediately prior to running simulation.
  gettimeofday(&start_tv, NULL);

  // Run simulation for # timesteps that user requested.
  for (i = 0; i < timesteps; i++)
  {
    // Calculate net force acting on each body.
    calculate_forces();

    // Move bodies appropriately based upon the net force acting on them.
    move_bodies();

    // Check for collisions and resolve any that are found.
    if (collisions_detected())
    {
      resolve_collisions();
    }

    // If print debug variable turned on, print to stdout.
    if (pdb) print_coordinates();

    // If GUI mode enabled write positions to file.
    if (enable_gui) export_positions();
  }

  // Record system's time immediately upon finishing simulation.
  gettimeofday(&end_tv, NULL);

  terminate();

  exit(EXIT_SUCCESS);
} /* END main() */


/* -----------------------------------------------------------------------------
 * Routine:     routine
 * Description: description
 * Arguments:   args
 * Returns:     retval
 * -------------------------------------------------------------------------- */

/* END sequential.c */
